### DISCUSSION TITLE
How to use body-parser as middleware?

### DISCUSSION THREAD

**Discussion by alex8bitw** [honojs/hono#2114] [Q&A]
Hi,

  I'd like to use the body-parser package as middleware in hono to parse c.req.body
I'm using the Fastly SDK which doesn't support c.req.parseBody() or .formData().

I'm trying to do  

import bodyParser, {BodyParser} from "body-parser";

app.use('*', bodyParser())

But I get an error.  Any idea how to extract Json form data from c.req.body when those two methods aren't available @ysugimoto ?
Thanks!

**Comment by ysugimoto** [comment]
@alex8bitw I'm not familiar with hono but hono seems to provide a request body parting method in `HonoRequest` so you don't need to use body-parser to get and extract json body I think.

```js
import { Hono } from "hono";
const app = new Hono();

app.post("/post", async (c) => {
  console.log(await c.req.parseBody()); // <- parse form request from application/x-www-form-urlencoded
  console.log(await c.req.json());            // <- parse json body from application/json
  return c.text("Hello, Fastly!");
});

app.fire();
```

The above code works in Fastly Compute, I hope this helps.

**Reply by alex8bitw** [reply]
Hi @ysugimoto , unfortunately, formData() and parseBody() (which uses formData internally) is not supported by the Fastly SDK,  It's unimplemented as per Fastly support.   I had already tried a version of the code above, and using both of the methods above, and they do not work (you get a TypeError).

**Reply by ysugimoto** [reply]
Understood. Certainly `c.req.parseBody()` raises an error in Fastly Compute.
I think the better way is that wait for Fastly's official implementation but we need to parse from the raw text body manually for now,
and unfortunately almost form-related npm modules cannot work in Fastly runtime.

```js
import { Hono } from "hono";
const app = new Hono();

app.post("/post", async (c) => {
  const body = await c.req.text();
  const formData = body
    .split("&")
    .reduce((prev, next) => {
      const kv = next.split("=").map(decodeURIComponent);
      return { ...prev, [kv[0]]: kv[1] };
    }, {});
  console.log(formData);
  return c.text("Hello, Fastly!");
});

app.fire();
```

Of course, this is an example and can't support `multipart/form-data`.

**Reply by alex8bitw** [reply]
I only need req.body.json() to be supported.  Are you saying "body-parser" (https://www.npmjs.com/package/body-parser)  won't work either?

**Comment by alex8bitw** [comment]
Ok, I've been able to get it working using let body = await c.req.json() and I do get back a proper json object.  Thank you!
